UI auto commands: UIEnter, UILeave

equalsraf opened the pull request:
Back in https://github.com/neovim/neovim/issues/3646 we never got into a consensus over this, but since gui_running is clearly gone maybe its time to push this forward again.

This PR adds 2 auto cmds

- ~~UIAttach~~ UIEnter
- ~~UIDetach~~ UILeave

I've avoided using the already existing GUIEnter and GUIFailed for fear of compatibility issues, but I'm fine with it either way.

- When attaching/detaching over api calls (nvim_ui_attach/detach) v:event holds the channel_id. When attaching/detaching the TUI (tui_bridge) v:event is empty.
- if channel_id (uint64) can not be represented as long (vimscript nr) then the autocmd is not triggered

equalsraf committed:
UIAttach, UIDetach

justinmk commented on src/nvim/ui.c:
```diff
@@ -220,11 +221,13 @@ void ui_busy_stop(void)
 void ui_attach_impl(UI *ui)
 {
   if (ui_count == MAX_UI_COUNT) {
+    apply_autocmds(EVENT_UIATTACHFAILED, NULL, NULL, FALSE, curbuf);
```
If this ever happens we will just raise the max. I don't think we need an event dedicated to it.

justinmk commented on src/nvim/api/ui.c:
```diff
@@ -112,6 +114,14 @@ void nvim_ui_attach(uint64_t channel_id, Integer width, Integer height,
+
+  if (channel_id <= LONG_MAX) {
+    dict_T *dict = get_vim_var_dict(VV_EVENT);
+    tv_dict_add_nr(dict, S_LEN("channel_id"), (long)channel_id);
```
Shouldn't it include all the other options as well?

bfredl commented on src/nvim/api/ui.c:
```diff
@@ -112,6 +114,14 @@ void nvim_ui_attach(uint64_t channel_id, Integer width, Integer height,
   pmap_put(uint64_t)(connected_uis, channel_id, ui);
   ui_attach_impl(ui);
+
+  if (channel_id <= LONG_MAX) {
```
this seems like an anti-pattern. Better to then add assert at channel/job creation that the number is not larger than VARNUMMER_MAX or what it is called.

justinmk replied:
Yep, this "should never" happen, and if it does we want a loud notice.

justinmk commented:
Should we name these ChannelOpen/Close in anticipation of the channel info PR? So we could merge this now and then enhance it later when @bfredl PR is done?

bfredl commented:
But these events do not describe the opening and closing of channels. They might be _typically_ be close in time, but that's something we don't assume, nor do I think we should enforce.

justinmk commented:
@bfredl Sure, I'm just thinking that it's good enough, and then we can later map it directly to channels. Maybe that is a bad idea, I'm not sure. I was thinking this might be a bridge (hack) that will solve the `gui_running` problem without adding events that will later become unnecessary (UIAttach/Detach).

bfredl commented:
@justinmk Maybe we should take a step back and think what this PR is for. The UI status/settings for an _individual_ channel I can fit into ChannelInfo later, but I suspect to replace/improve gui_running one wants an autocmd for _global_ ui status, i e 

- The _final_ calculated screen size and set of activated widgets
- Whether _any_ UI is remote (or perhaps the number)
- Whether _any_ UI uses gui colors and term colors respectively

equalsraf commented:
> The final calculated screen size and set of activated widgets

You mean external widgets right? like the pum, etc

> but I suspect to replace/improve gui_running one wants an autocmd for global ui status

If the idea is to provide general information about all available UIs then perhaps its better to simplify this PR by removing the data from v:event (leaving the channel id) and provide an api to get UI info in general. Something like `nvim_list_ui()` and `nvim_ui_get_option()`.

I'm good either way. Its mostly an issue of figuring out what is the proper (expected) way to handle this kind of thing. I got started on this after https://github.com/equalsraf/neovim-qt/pull/323 because people wanted to hook up stuff to GUI attachment/detach.

justinmk commented on runtime/doc/autocmd.txt:
```diff
@@ -263,8 +263,10 @@ Name			triggered by ~
-|GUIEnter|		after starting the GUI successfully
-|GUIFailed|		after starting the GUI failed
+|GUIEnter|		never used in Nvim.
+|GUIFailed|		never used in Nvim.
```
should be moved to `deprecated.txt` I think.

justinmk commented on runtime/doc/autocmd.txt:
```diff
@@ -329,6 +331,9 @@ Name			triggered by ~
 |CompleteDone|		after Insert mode completion is done
 
+|UIAttach|		after a UI attaches successfully
+|UIDetach|		after a UI detaches
```
looks redundant with the ones above

justinmk commented:
@equalsraf I was thinking similar. But @bfredl is right about the "final calcuated UI". Though I'm not sure it makes sense to put that in the UIAttach event info. 

It seems like some thing will be awkward without UIAttach/Detach, so let's do the bare minimum to make them available for 0.2.1.

justinmk commented on src/nvim/ui_bridge.c:
```diff
@@ -88,6 +89,7 @@ UI *ui_bridge_attach(UI *ui, ui_main_fn ui_main, event_scheduler scheduler)
   uv_mutex_unlock(&rv->mutex);
 
   ui_attach_impl(&rv->bridge);
+  apply_autocmds(EVENT_UIATTACH, NULL, NULL, false, curbuf);
```
v:event should be available here too.

equalsraf replied:
I assume this no longer applies since channel_id is only available in external attachment.

justinmk replied:
Even so, `v:event` should always be available for a given event type, even if it's something like `v:event.channel_id = v:null`.

It would be weird if `v:event` were available for one UIAttach invocation but not another.

bfredl replied:
`chanid = 0` would be the canonical sentinel value for "no channel" .

justinmk replied:
@bfredl But we use `0` to mean "broadcast", so maybe it should be `-1` or `v:null`.

bfredl replied:
Broadcast means "no specific channel". We also use 0 when connection/job failed, and thus no channel returned. Using different sentinels for "no channel" because the alternative is different is more confusing and surprising than helpful

justinmk replied:
Ok.

equalsraf replied:
Pushing changes for this one. Any idea how to test this? I cant figure out how to catch this autocommand.

justinmk commented on src/nvim/api/ui.c:
```diff
@@ -112,6 +114,32 @@ void nvim_ui_attach(uint64_t channel_id, Integer width, Integer height,
+
+  tv_dict_add_dict(dict, S_LEN("options"), options_vim);
+  tv_dict_set_keys_readonly(dict);
+  apply_autocmds(EVENT_UIATTACH, NULL, NULL, false, curbuf);
```
`apply_autocmds` should be queued on the main loop, rather than immediately executed.

    multiqueue_put(main_loop.events ...

bfredl replied:
No, this is already on the main loop and safe.

justinmk replied:
I'm not concerned about threads. when new events are added, executing VimL in new codepaths is often buggy. It happened with WinNew. There's no reason to execute this code immediately here, it's better to defer to the next tick.

bfredl replied:
I never said anything about threads. This _already_ is in a handler of a _deferred_ event. Absolutely nothing is gaining by "deferring to the next tick" because we already is one tick deferred from whatever dangerous place you want to avoid.

justinmk replied:
Rather than safety, I wonder if we should avoid the risk of slow plugins potentially slowing-down UI attach.

MyXelf commented:
Is there anything that can be done to commit this? Sorry if I'm missing something obvious here, but I've been moving back and forward with downstream (neovim-qt) and upstream (neovim) seeing a lot of related efforts in equalsraf/neovim-qt#323, equalsraf/neovim-qt#156, equalsraf/neovim-qt#158 and of course this thread.

equalsraf commented:
@MyXelf I'm the one delaying this. All the comments in the earlier discussion need to be addressed, and I lack the time. https://github.com/neovim/neovim/pull/6917#issuecomment-338390906 is probably the hardest bit.

I'm pretty sure I wont finish this during the week, so if you want to take over this PR I have no objections :)

bfredl commented on src/nvim/api/ui.c:
```diff
@@ -160,6 +168,12 @@ void nvim_ui_detach(uint64_t channel_id, Error *err)
   remote_ui_disconnect(channel_id);
+
+  dict_T *dict = get_vim_var_dict(VV_EVENT);
+  tv_dict_add_nr(dict, S_LEN("channel_id"), (long)channel_id);
```
`nvim_list_uis()` uses `chan`, perhaps the same here for consistency.

equalsraf replied:
Ok renamed to chan

bfredl commented on test/functional/api/ui_spec.lua:
```diff
@@ -34,4 +35,13 @@ describe('nvim_ui_attach()', function()
   end)
+  it('autocmds UIAttach/Detach set v:event', function()
+    local screen = Screen.new()
+    command('autocmd UIAttach * :let g:ui_attach_v_event_has_channel_id = has_key(v:event, "channel_id")')
```
you can store the actual channel id, we know it to be `1` for the stdio channel used by tests.

equalsraf replied:
Changed this to compare the full dict

bfredl commented on test/functional/api/ui_spec.lua:
```diff
@@ -34,4 +35,13 @@ describe('nvim_ui_attach()', function()
+    command('autocmd UIAttach * :let g:ui_attach_v_event_has_channel_id = has_key(v:event, "channel_id")')
+    command('autocmd UIDetach * :let g:ui_detach_v_event_has_channel_id = has_key(v:event, "channel_id")')
+    screen:attach()
+    screen:detach()
```
put this after the first assert?

equalsraf replied:
Moved.

justinmk committed:
UIAttach, UIDetach

doc: ginit.vim, gvimrc
fix #3656

justinmk commented:
Updated. As discussed above, since we have `nvim_list_uis()` there is not much need for `v:event` to have anything but the channel id. Though we can add more fields later if it's too clumsy.

justinmk commented on src/nvim/ui.c:
```diff
@@ -292,9 +293,13 @@ void ui_attach_impl(UI *ui)
   }
   ui_refresh();
+
+  if (!comp) {
```
@bfredl is there a more clever way to check if the UI is the compositor? `ui_comp_init` happens before eval is available, also I guess we don't want an extra UIAttach event just for the compositor. 

see also https://github.com/neovim/neovim/issues/10937#issuecomment-528236541

bfredl replied:
The first UI that gets added `uis[0]` is always the compositor.

justinmk committed:
API/nvim_list_uis(): include "chan" field for TUI

Even though it's always zero currently, it's less confusing if all UIs
have the same fields.

justinmk committed:
rename: UIAttach/UIDetach => UIEnter/UILeave

"enter"/"leave" is more conventional for Vim events, and
"attach"/"detach" distinction does not gain much.

justinmk committed:
UIEnter/UILeave: fire for embedder UI, builtin TUI

Before this, --embed UIs (without --headless) would not trigger UIEnter.

For TUI, maybe UIEnter isn't useful, but:
- It is less "surprising"/special.
- Makes documentation simpler.
- When TUI becomes a coprocess, it will happen anyway.
